### PR TITLE
[RISCV] Use subreg extract for extract_vector_elt when vlen is known

### PR DESCRIPTION
**Pull request by preames**
This is the first in a planned patch series to teach our vector lowering how to exploit register boundaries in LMUL>1 types when VLEN is known to be an exact constant.  This corresponds to code compiled by clang with the -mrvv-vector-bits=zvl option.

For extract_vector_elt, if we have a constant index and a known vlen, then we can identify which register out of a register group is being accessed.  Given this, we can do a sub-register extract for that register, and then shift any remaining index.

This results in all constant index extracts becoming m1 operations, and thus eliminates the complexity concern for explode-vector idioms at high lmul.

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 1]
```diff
@@ -7895,6 +7895,29 @@ SDValue RISCVTargetLowering::lowerEXTRACT_VECTOR_ELT(SDValue Op,
...
+      VecVT.getSizeInBits().getKnownMinValue() > MinVLen) {
+    unsigned OrigIdx = IdxC->getZExtValue();
+    EVT ElemVT = VecVT.getVectorElementType();
+    unsigned ElemSize = ElemVT.getSizeInBits().getKnownMinValue();
+    unsigned ElemsPerVReg = MinVLen / ElemSize;
+    unsigned RemIdx = OrigIdx % ElemsPerVReg;
+    unsigned SubRegIdx = OrigIdx / ElemsPerVReg;
+    unsigned ExtractIdx =
+      SubRegIdx * ContainerVT.getVectorElementCount().getKnownMinValue();
+    ContainerVT = getLMUL1VT(ContainerVT);
```
Comment (lukel97) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:7915]: Does the ExtractIdx not need to be computed in terms of the LMUL1 ContainerVT?

Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp:7915]: I think it does. I checked out this patch and all of the tests hit an assert with the current code.

Comment (preames) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp:7915]: This is weird.  I know exactly what the problem is here; I moved the redefinition of ContainerVT below the ElementIdx initialization.  The weird part is that I swear I saw that bug downstream, and fixed it before pushing this branch for review.  And yet, the broken code is both in the remote and local branches.  All I can think is that I screwed up a rebase and lost a change.  

Sorry for the noise here, will republish correct patch once back from holiday.

Comment (preames) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp]: Fixed version pushed, sorry again for the noise.

**Review by topperc — APPROVED** [review]
LGTM

**Pull request merged**